### PR TITLE
checkers: add badCond checker

### PR DESCRIPTION
**Pull request by Quasilyte**
Signed-off-by: Iskander Sharipov <quasilyte@gmail.com>

### CODE REVIEW

File: checkers/internal/lintutil/astflow.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+//
+// If proven really useful, can be moved to go-toolsmith library.
+
+// IsImmutable reports whether n can be midified through any operation.
+func IsImmutable(info *types.Info, n ast.Expr) bool {
+	if astp.IsBasicLit(n) {
+		return true
+	}
+	tv, ok := info.Types[n]
+	return ok && !tv.Assignable() && !tv.Assignable()
```
Comment (cristaloleg) [inline checkers/internal/lintutil/astflow.go]:
```suggestion
	return ok && !tv.Assignable()
```

File: checkers/testdata/badCond/positive_tests.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package checker_test
+
+func bad1(retVal []int, start int) {
+	/*! `i > start` in loop; probably meant `i < start`? */
+	for i := 0; i > start; i++ {
+		retVal[i] = 0
+	}
+}
+
```
Comment (cristaloleg) [inline checkers/testdata/badCond/positive_tests.go]:
```suggestion

var i int
for ; i > start; i++ {
}

```

File: checkers/internal/lintutil/astflow.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	if dst, ok := dst.(*ast.Ident); ok {
+		// Identifier can be shadowed,
+		// so we need to check the object as well.
+		obj := info.ObjectOf(dst)
+		if obj == nil {
+			return true // Being conservative
+		}
+		isDst = func(x ast.Expr) bool {
+			if x, ok := x.(*ast.Ident); ok {
+				return x.Name == dst.Name && info.ObjectOf(x) == obj
```
Comment (cristaloleg) [inline checkers/internal/lintutil/astflow.go]:
```suggestion
				return ok && x.Name == dst.Name && info.ObjectOf(x) == obj
				
```

And drop `return false`

File: checkers/internal/lintutil/astflow.go (hunk 1 above)
Comment (Quasilyte) [reply checkers/internal/lintutil/astflow.go]: Oh, it should be `!tv.Assignable() && !tv.Addressible()`

File: checkers/testdata/badCond/positive_tests.go (hunk 2 above)
Comment (Quasilyte) [reply checkers/testdata/badCond/positive_tests.go]: You can add handling for this one. :)

File: checkers/internal/lintutil/astflow.go (hunk 3 above)
Comment (Quasilyte) [reply checkers/internal/lintutil/astflow.go]: Fixed, thanks.

**Pull request merged**